File: 121.best_time_to_buy_and_sell_stock/maxProfit.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    if(prices.length === 0 || prices.length === 1) return 0;
    var maxNum = prices[prices.length-1];
    var maxValue = 0;
    for(var i = prices.length-2; i >=0; i--){
        maxValue = Math.max(maxValue,(maxNum - prices[i]));
        maxNum = Math.max(maxNum, prices[i]);
    }
    return maxValue;
};

console.log(maxProfit([7, 1, 5, 3, 6, 4]));
console.log(maxProfit([7, 6, 4, 3, 1]));

File: 121.best_time_to_buy_and_sell_stock/maxProfit.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        profit, res = 0, 0
        dif = []

        for i in xrange(1,len(prices)):
            dif.append(prices[i] - prices[i-1])

        for i in xrange(len(dif)):
            profit += dif[i]
            if profit < 0:
                profit = 0
            res = max(profit, res)
        return res
